lib/utils/ignore.ts

```diff
@@ -0,0 +1,151 @@
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+
+import dockerIgnore = require('@zeit/dockerignore');
+import ignore from 'ignore';
+
+export enum IgnoreFileType {
+	DockerIgnore,
+	GitIgnore,
+}
+
+interface IgnoreEntry {
+	pattern: string;
+	// The relative file path from the base path of the build context
+	filePath: string;
+}
+
+export class FileIgnorer {
+	private dockerIgnoreEntries: IgnoreEntry[];
+	private gitIgnoreEntries: IgnoreEntry[];
+
+	private static ignoreFiles: Array<{
+		pattern: string;
+		type: IgnoreFileType;
+		allowSubdirs: boolean;
+	}> = [
+		{
+			pattern: '.gitignore',
+			type: IgnoreFileType.GitIgnore,
+			allowSubdirs: true,
+		},
+		{
+			pattern: '.dockerignore',
+			type: IgnoreFileType.DockerIgnore,
+			allowSubdirs: false,
+		},
+	];
+
+	public constructor(public basePath: string) {
+		this.dockerIgnoreEntries = [];
+		this.gitIgnoreEntries = [];
+	}
+	/**
+	 * @param  {string} relativePath
+	 * 	The relative pathname from the build context, for example a root level .gitignore should be
+	 * 		./.gitignore
+	 * @returns IgnoreFileType
+	 * 	The type of ignore file, or null
+	 */
+	public isIgnoreFile(relativePath: string): IgnoreFileType | null {
+		for (const { pattern, type, allowSubdirs } of FileIgnorer.ignoreFiles) {
+			if (relativePath === pattern || relativePath === `./${pattern}`) {
+				return type;
+			}
+			const basename = path.posix.basename(relativePath);
+			if (allowSubdirs && basename === pattern) {
+				return type;
+			}
+		}
+
+		return null;
+	}
+	/**
+	 * @param  {string} fullPath
+	 * 	The full path on disk of the ignore file
+	 * @param  {IgnoreFileType} type
+	 * @returns Promise
+	 */
+	public async addIgnoreFile(
+		fullPath: string,
+		type: IgnoreFileType,
+	): Promise<void> {
+		const contents = await fs.readFile(fullPath, 'utf8');
+
+		contents.split('\n').forEach(line => {
+			const trimmed = line.trimLeft();
```

Comment:
Why do you only trim the left side?  I know you do the left side at a minimum for comment purposes but it seems like either both sides or neither would be necessary (how do you ignore a file with leading/trailing spaces in the name?)

Reply:
Yeah you're right, the ignore files do take into account leading and trailing whitespace, so I just need to use the trimmed version to check for comments, and the actual version for the ignore pattern - good catch!

Reply:
In that case it's probably simpler to do
```typescript
if (/\s*#/.test(line)) {
	return;
}
```
and avoid needing to create a copy of the string